nvflare/fuel/f3/drivers/aio_grpc_driver.py

```diff
@@ -0,0 +1,412 @@
+                conn_ctx.error = f"failed to start server: {type(ex)}: {ex}"
+        conn_ctx.waiter.set()
+
+    def listen(self, connector: ConnectorInfo):
+        self.logger.debug(f"listen called from thread {threading.current_thread().name}")
+        self.connector = connector
+        aio_ctx = AioContext.get_global_context()
+        conn_ctx = _ConnCtx()
+        aio_ctx.run_coro(self._start_server(connector, aio_ctx, conn_ctx))
+        while not conn_ctx.conn and not conn_ctx.error:
+            self.logger.debug("SERVER: waiting for server to be started")
+            time.sleep(0.1)
+        if conn_ctx.error:
+            raise CommError(code=CommError.ERROR, message=conn_ctx.error)
+        self.logger.debug("SERVER: waiting for server to finish")
+        conn_ctx.waiter.wait()
+        self.logger.debug("SERVER: server is done")
+
+    async def _start_connect(self, connector: ConnectorInfo, aio_ctx: AioContext, conn_ctx: _ConnCtx):
+        self.logger.debug("Started _start_connect coro")
+        self.connector = connector
+        params = connector.params
+        address = get_address(params)
+
+        self.logger.debug(f"CLIENT: trying to connect {address}")
+        try:
+            secure = ssl_required(params)
+            if secure:
+                grpc_channel = grpc.aio.secure_channel(
+                    address, options=self.options, credentials=self.get_grpc_client_credentials(params)
+                )
+            else:
+                grpc_channel = grpc.aio.insecure_channel(address, options=self.options)
+
+            async with grpc_channel as channel:
+                self.logger.debug(f"CLIENT: connected to {address}")
+                stub = StreamerStub(channel)
+                self.logger.debug("CLIENT: got stub!")
+                conn_props = {DriverParams.PEER_ADDR.value: address}
+
+                if secure:
+                    conn_props[DriverParams.PEER_CN.value] = "N/A"
+
+                connection = AioStreamSession(
+                    side="CLIENT", aio_ctx=aio_ctx, connector=connector, conn_props=conn_props, channel=channel
+                )
+
+                try:
+                    self.logger.debug(f"CLIENT: start streaming on connection {connection}")
+                    msg_iter = stub.Stream(connection.generate_output())
+                    conn_ctx.conn = connection
+                    await connection.read_loop(msg_iter)
+                except asyncio.CancelledError as error:
+                    self.logger.debug(f"CLIENT: RPC cancelled: {error}")
+                except BaseException as ex:
+                    if self.closing:
+                        self.logger.debug(f"Connection {connection} closed by {type(ex)}: {ex}")
+                    else:
+                        self.logger.error(f"Connection {connection} client read exception {type(ex)}: {ex}")
+                    self.logger.debug(traceback.format_exc())
```

Comment:
for exception and traceback please use things in: https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/security/logging.py

Reply:
We should do this all at once when everything is merged in.